From 1a22111ce748eed477f60ad9bbb3434b1f60d0a3 Mon Sep 17 00:00:00 2001
From: David Hildenbrand <dahi@linux.vnet.ibm.com>
Date: Mon, 11 May 2015 17:52:14 +0200
Subject: [PATCH 3/9] sched/preempt, futex: Disable preemption in UP
 futex_atomic_op_inuser() explicitly

commit feb3b8e77e3d0567bd2c25afee0fac0c7e7f377d upstream

Let's explicitly disable/enable preemption in the !CONFIG_SMP version
of futex_atomic_cmpxchg_inatomic, to prepare for pagefault_disable() not
touching preemption anymore. This is needed for this function to be
callable from both, atomic and non-atomic context.

Otherwise we might break mutual exclusion when relying on a get_user()/
put_user() implementation.

[upstream commit f3dae07e442a8131a5485b6a38db2aa22a7a48cf]
Signed-off-by: David Hildenbrand <dahi@linux.vnet.ibm.com>
Signed-off-by: Alexandru Moise <alexandru.moise@windriver.com>
---
 include/asm-generic/futex.h |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/include/asm-generic/futex.h b/include/asm-generic/futex.h
index b59b5a5..b697fe3 100644
--- a/include/asm-generic/futex.h
+++ b/include/asm-generic/futex.h
@@ -106,6 +106,7 @@ futex_atomic_cmpxchg_inatomic(u32 *uval, u32 __user *uaddr,
 {
 	u32 val;
 
+	preempt_disable();
 	if (unlikely(get_user(val, uaddr) != 0))
 		return -EFAULT;
 
@@ -113,6 +114,7 @@ futex_atomic_cmpxchg_inatomic(u32 *uval, u32 __user *uaddr,
 		return -EFAULT;
 
 	*uval = val;
+	preempt_enable();
 
 	return 0;
 }
-- 
1.7.5.4

